docs: Use Solidity lexer from PyPI, instead of a local module

veox opened the pull request:
Mentioned in https://github.com/ethereum/solidity/issues/3637#issuecomment-390983359.

This PR:

* switches syntax highlighting from a local Pygments lexer to one pulled from PyPI;
* removes the aforementioned local lexer from the source tree;
* changes the CircleCI doc-building job to use the RTD theme.

Before PR: https://14123-40892817-gh.circle-artifacts.com/0/docs-html/index.html (thanks Chris!)

After PR: https://14305-40892817-gh.circle-artifacts.com/0/docs-html/index.html (at ad975a6)

-----

The [lexer package](https://pypi.org/project/pygments-lexer-solidity/) is maintaned by me, [repo on gitlab](https://gitlab.com/veox/pygments-lexer-solidity).

veox committed:
docs: use SolidityLexer from pygments_lexer_solidity package.

veox committed:
docs: remove utils/SolidityLexer.py

Note that some features present in this local module have not been
included in the upstream package yet!..

veox committed:
docs: install requirements as part of doc-build script.

veox committed:
circleci: install python-pip as part of doc build dependency.

Required to install the lexer package.

chriseth commented:
Documentation generated via CI is here: https://14175-40892817-gh.circle-artifacts.com/0/docs-html/index.html

Current develop for comparison: https://14123-40892817-gh.circle-artifacts.com/0/docs-html/index.html

Note that these are different from the "read the docs" documentation because they use a different theme.

chriseth commented:
A quick inspection showed that this one does not hilight natspec comments. I did not particularly like the way natspec was highlighted, but we should do it in some way.

Also, does this theme hilight all `.sender` or just `msg.sender`?

In general, I like it, though!

veox commented:
> A quick inspection showed that this one does not hilight natspec comments. I did not particularly like the way natspec was highlighted, but we should do it in some way.

Yes, the stand-alone lexer package doesn't do NatSpec (yet). EDIT: I've now got [a working prototype](https://gitlab.com/veox/pygments-lexer-solidity/commit/582a1c6f755d68e037765084fa41adf163c25673) that does this. EDIT2: Now used since commit 4004fa8.

> Also, does this theme hilight all .sender or just msg.sender?

~~It~~ The lexer [hightlights all uses of`sender`](https://gitlab.com/veox/pygments-lexer-solidity/blob/master/pygments_lexer_solidity/lexer.py#L213).

This should be relatively simple to change; is there a place that lists all possible ~~uses~~ permutations of ~~such~~ these nested keywords?..

EDIT: found in [the "global variables" section](https://solidity.readthedocs.io/en/v0.4.24/miscellaneous.html#global-variables). EDIT2: observed in `pygments-lexer-solidity` v0.3.1 - used since commit ad975a6.

veox commented:
> Note that these are different from the "read the docs" documentation because they use a different theme.

The theme should now be installed as part of the CI process; should I add an option to the config, so that RTD's theme is chosen instead?

EDIT: I've added this in 3678450 - revert if not desirable.

veox commented:
The current development version of `pygments-lexer-solidity` does NatSpec like this (with the `default` theme, for me locally):

![](https://veox.pw/dump/natspec.png)

Does this look acceptable? If so, I can do a release, and push to this PR's branch to use the newer version.

EDIT: done in 4004fa8.

veox committed:
docs: bump required Solidity lexer package version.

pygments-lexer-solidity 0.3.0 highlights NatSpec specials like
@author, @title, etc..

veox committed:
docs: change default theme to RTD's.

RTD probably has their `default` theme set to their own `rtd` theme,
while CircleCI has some regular default Sphinx theme.

This changes the Sphinx project configuration to use `rtd` as the
theme.

veox committed:
docs: bump Solidity lexer package to v0.3.1.

That release includes stricter matching of "nested" globals, i.e.
it highlights `msg.sender` but not `sender`.

axic commented:
Looks like a great idea!

@veox one question though, what you think how long will it take to keep this up to date with 0.5.0? Is there perhaps a way to have an option choose the "major" version of Solidity?

axic commented on scripts/docs.sh:
```diff
@@ -28,5 +28,6 @@
 
 set -e
 cd docs
+pip install -r requirements.txt
```
What is this file?

axic commented on docs/conf.py:
```diff
@@ -112,7 +112,7 @@ def setup(sphinx):
 # The theme to use for HTML and HTML Help pages.  See the documentation for
 # a list of builtin themes.
-html_theme = 'default'
+html_theme = 'sphinx_rtd_theme'
```
Oh nice, so as far as I understand this ensures that locally generated style is the same as on RTD?

veox replied:
Yes, for local builds and on Circle CI. That does mean the RTD theme has to be installed.

veox commented:
@axic The last time I reviewed Solidity keywords was after 0.4.22 - so I think all the current keywords should be there?..

Sub-states _could_ be maintained in the lexer, determining the keywords-to-be-highlighted based on `pragma solidity` - but that would add some complexity, and I don't yet see the advantage.

EDIT: It _might_ make sense to make this distinction with Solidity 1.0, when the keywords will (hopefully) change less often than currently.

chriseth commented:
@veox the "actual" user-natspec comments work without tag - are they also highlighted in some way?
```
/// does something
function f() { ... }
```

veox commented:
@chriseth The lexer has absolutely no differentiation of `//` (regular comment) and `///` (natspec); it just highlights several `@`-prefixed keywords when they appear in a `comment` of any sort.

In other words, it's all "just `comments`" as far as the lexer is concerned.

-----

OT: Actual NatSpec parsing would have to be done by a different tool, Pygments is not well-suited for this.
